File: exercism/javascript/nucleotide-count/nucleotide-count.js
import { it, _ } from "param.macro"
import { adjust, findIndex, tap, split, reduce, join } from "ramda"

const nucleotides = ["A", "C", "G", "T"]

const guardInvalid = tap(
  it === -1 && throw new Error("Invalid nucleotide in strand")
)

const count = (counts, nucleotide) =>
  nucleotides
  |> findIndex(it === nucleotide)
  |> guardInvalid
  |> adjust(_, it + 1, counts)

export const parse = sequence =>
  sequence |> split("") |> reduce(count, [0, 0, 0, 0]) |> join(" ")

File: exercism/clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count
  (:require [clojure.core :as core]))

(def nucleotides [\A \T \C \G])

(defn nucleotide? [nucleotide]
  (.contains nucleotides nucleotide))

(defn count [nucleotide strand]
  {:pre [(nucleotide? nucleotide)]}
  (->> strand
       (filter #(= % nucleotide))
       (core/count)))

(defn nucleotide-counts [strand]
  {\A (count \A strand)
   \T (count \T strand)
   \C (count \C strand)
   \G (count \G strand)})
